### PR TITLE
Fixes and improvements for image build and deploy mojo and tasks

### LINKED ISSUE #32226: mvn quarkus:deploy fails on a when quarkus-openshift is added to the project
### Describe the bug

In a project that contains `quarkus-openshift` I am running `mvn quarkus:deploy`. The execution fails with an error about both docker and openshift contianer image extension being present.

### Expected behavior

A single container image extension is available and therefore the command should work.

### Actual behavior

[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:deploy (default-cli) on project demo: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.container.image.deployment.ContainerImageProcessor#publishImageInfo threw an exception: java.lang.IllegalStateException: quarkus-container-image-openshift and quarkus-container-image-docker were detected, at most one container-image extension can be present.
[ERROR] Either remove the unneeded ones, or select one by setting any of the following configuration properties: 'quarkus.container-image.builder=docker', 'quarkus.container-image.builder=openshift'
[ERROR] 	at io.quarkus.container.image.deployment.ContainerImageCapabilitiesUtil.getActiveContainerImageCapability(ContainerImageCapabilitiesUtil.java:40)
[ERROR] 	at io.quarkus.container.image.deployment.ContainerImageProcessor.ensureSingleContainerImageExtension(ContainerImageProcessor.java:112)
[ERROR] 	at io.quarkus.container.image.deployment.ContainerImageProcessor.publishImageInfo(ContainerImageProcessor.java:54)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:833)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] -> [Help 1]

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by iocanel**
Resolves: #32226

Also addresses a ton of issues spotted with I was on it:

- Deploy tasks was competely broken
- Both image build & deploy tasks/mojo didn't recognize `quarkus-openshift` as a container image extension.
- Both deploy tasks/mojo were getting confused by the presence of image builders in the project.
- A lot of code dublication.

### CODE REVIEW

**Comment by iocanel** [comment]
I don't think that the failures are related as these seem to standard errors these days.

File: devtools/maven/src/main/java/io/quarkus/maven/ImageBuilder.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                .or(() -> projectBuilders.stream().findFirst());
+    }
+
+    /**
+     * Get teh image builder extensions found in the project.
+     *
+     * @param the project to search for extensions
+     * @return A set with the discovered extenions.
+     */
+    public static Set<String> getProjecBuilder(MavenProject project) {
```
Comment (Sgitario) [inline devtools/maven/src/main/java/io/quarkus/maven/ImageBuilder.java:73]: There is a typo here.

**Review by Sgitario — CHANGES_REQUESTED** [review]
Because of the refactoring changes, I can't easily see the actual fix, perhaps what fixes the linked issue is that you're filtering out the transitive dependencies?

Also, I tried these changes and `mvn quarkus:deploy` now does not fail, but the OpenShift S2i deployment is not triggered (using `mvn clean install -Dquarkus.kubernetes.deploy=true` does work). I expected that `mvn quarkus:deploy` was similar to `mvn clean install -Dquarkus.kubernetes.deploy=true`.

**Comment by iocanel** [comment]
Just realized that this pull request hasn't been merged yet and I have already created a follow up request #32410 that does includes all the commitss of this one.

**Pull request closed without merging**